File: web/footprint.js
// Footprint cells — same geometry as internal/chart/footprint.go.
// Left = sell (SideAsk hit the bid), right = buy (SideBid lifted ask).
// Gold stroke is imbalance; thicker is stacked. Opacity is volume
// versus the max cell in that bar.
(function () {
  const up = "#1b7f3a";
  const down = "#b42318";
  const imb = "#eab308";

  function cellMark(imbs, px, dir) {
    imbs = imbs || [];
    for (let i = 0; i < imbs.length; i++) {
      if (imbs[i].Price === px && imbs[i].Dir === dir) {
        return { hit: true, stacked: !!imbs[i].Stacked };
      }
    }
    return { hit: false, stacked: false };
  }

  function draw(ctx, box, vis, yOf, fp) {
    if (!fp || !fp.Bars) return;
    const step = fp.TicksPerRow > 0 ? fp.TicksPerRow : 4;
    const h = Math.max(1, Math.abs(yOf(0) - yOf(step)));
    const slot = vis.n <= 0 ? 0 : (box.right - box.left) / vis.n;
    const half = Math.max(1, slot * 0.4);
    const n = Math.min(fp.Bars.length, vis.i1);
    for (let i = vis.i0; i < n; i++) {
      const bar = fp.Bars[i];
      if (!bar || !bar.Levels) continue;
      let maxV = 0;
      for (let k = 0; k < bar.Levels.length; k++) {
        if (bar.Levels[k].Buy > maxV) maxV = bar.Levels[k].Buy;
        if (bar.Levels[k].Sell > maxV) maxV = bar.Levels[k].Sell;
      }
      if (maxV <= 0) continue;
      const x = box.left + (i - vis.i0 + 0.5) * slot;
      for (let k = 0; k < bar.Levels.length; k++) {
        const lv = bar.Levels[k];
        const y = yOf(lv.Price) - h / 2;
        if (lv.Sell > 0) {
          const m = cellMark(bar.Imbs, lv.Price, 1);
          cell(ctx, x - half, y, half, h, down, lv.Sell, maxV, m);
        }
        if (lv.Buy > 0) {
          const m = cellMark(bar.Imbs, lv.Price, 2);
          cell(ctx, x, y, half, h, up, lv.Buy, maxV, m);
        }
      }
    }
  }

  function cell(ctx, x, y, w, h, color, vol, max, mark) {
    ctx.fillStyle = color;
    ctx.globalAlpha = 0.20 + 0.70 * vol / max;
    ctx.fillRect(x, y, w, h);
    ctx.globalAlpha = 1;
    if (mark.hit) {
      ctx.strokeStyle = imb;
      ctx.lineWidth = mark.stacked ? 2.5 : 1.5;
      ctx.strokeRect(x + 0.5, y + 0.5, Math.max(0, w - 1), Math.max(0, h - 1));
    }
  }

  function at(fp, barIndex, price) {
    if (!fp || !fp.Bars || barIndex < 0 || barIndex >= fp.Bars.length) {
      return { buy: 0, sell: 0 };
    }
    const bar = fp.Bars[barIndex];
    const step = fp.TicksPerRow > 0 ? fp.TicksPerRow : 4;
    const row = price - (price % step);
    let buy = 0;
    let sell = 0;
    (bar.Levels || []).forEach(function (lv) {
      if (lv.Price === row) {
        buy = lv.Buy;
        sell = lv.Sell;
      }
    });
    return { buy: buy, sell: sell };
  }

  window.MDL = window.MDL || {};
  window.MDL.footprint = { draw: draw, at: at };
})();

File: internal/chart/footprint.go
package chart

import (
	"fmt"
	"io"
	"math"

	"github.com/morteza-sakifard/TickEngine/internal/core"
)

const imbColor = "#eab308"

func hasFootprint(v View) bool {
	return v.Footprint != nil && len(v.Footprint.Bars) > 0
}

func writeFootprint(w io.Writer, fp *FootprintView, sc Scale) {
	if fp == nil {
		return
	}
	step := fp.TicksPerRow
	if step < 1 {
		step = 1
	}
	h := math.Abs(sc.Y(0) - sc.Y(core.Ticks(step)))
	if h < 1 {
		h = 1
	}
	n := len(fp.Bars)
	if n > sc.N {
		n = sc.N
	}
	half := sc.slotWidth() * 0.4
	if half < 1 {
		half = 1
	}
	for i := 0; i < n; i++ {
		bar := fp.Bars[i]
		var maxV core.Qty
		for _, lv := range bar.Levels {
			if lv.Buy > maxV {
				maxV = lv.Buy
			}
			if lv.Sell > maxV {
				maxV = lv.Sell
			}
		}
		if maxV <= 0 {
			continue
		}
		x := sc.X(i)
		for _, lv := range bar.Levels {
			y := sc.Y(lv.Price) - h/2
			if lv.Sell > 0 {
				st, imb := cellMark(bar.Imbs, lv.Price, core.SideBid)
				writeFootCell(w, x-half, y, half, h, downColor, lv.Sell, maxV, st, imb)
			}
			if lv.Buy > 0 {
				st, imb := cellMark(bar.Imbs, lv.Price, core.SideAsk)
				writeFootCell(w, x, y, half, h, upColor, lv.Buy, maxV, st, imb)
			}
		}
	}
}

func writeFootCell(w io.Writer, x, y, width, height float64, color string, vol, max core.Qty, stacked bool, imb bool) {
	op := 0.20 + 0.70*float64(vol)/float64(max)
	fmt.Fprintf(w, "  <rect x=\"%.1f\" y=\"%.1f\" width=\"%.1f\" height=\"%.1f\" fill=\"%s\" fill-opacity=\"%.2f\"",
		x, y, width, height, color, op)
	if imb {
		sw := 1.5
		if stacked {
			sw = 2.5
		}
		fmt.Fprintf(w, " stroke=\"%s\" stroke-width=\"%.1f\"", imbColor, sw)
	}
	io.WriteString(w, "/>\n")
}

func cellMark(imbs []FootprintImb, px core.Ticks, dir core.Side) (stacked bool, imb bool) {
	for _, im := range imbs {
		if im.Price == px && im.Dir == dir {
			return im.Stacked, true
		}
	}
	return false, false
}
